libtrap/src/ifc_tcpip.c

```diff
@@ -960,419 +958,533 @@ static int client_socket_connect(void *priv, const char *dest_addr, const char *
+         case 0:
+            /* Succesfully locked, buffer can be used */
+            break;
+         case ETIMEDOUT:
+            /* Desired buffer is still full after timeout */
+            if (block) {
+               /* Blocking send, wait until buffer is free to use */
+               pthread_mutex_unlock(&c->ctx->out_ifc_list[c->ifc_idx].ifc_mtx);
+               goto repeat;
+            } else {
+               /* Non-blocking send, drop message or force buffer reset (not implemented) */
+               goto timeout;
             }
-         }
+         default:
+            VERBOSE(CL_ERROR, "Unexpected error in pthread_mutex_timedlock()");
+            goto timeout;
       }
-      if (passed == c->connected_clients) {
-         /* there is no client that failed */
-         for (i = 0, j = 0; i < c->clients_arr_size; ++i) {
-            cl = &c->clients[i];
-            if ((cl->sd > 0) && (cl->client_state == CURRENT_COMPLETE)) {
-               cl->client_state = CURRENT_IDLE;
-               j++;
-               if (j == passed) {
-                  break;
-               }
-            }
-         }
-         result = TRAP_E_OK;
-      } else {
-         if (block != 0) {
-            goto blocking_repeat;
-         }
+   }
+
+   /* Check if there is enough space in buffer */
+   free_bytes = c->buffer_size - buffer->wr_index;
+   if (free_bytes >= (size + sizeof(size))) {
+      /* Store message into buffer */
+      insert_into_buffer(buffer, data, size);
+
+      /* If bufferswitch is 0, only 1 message is allowed to be stored in buffer */
+      if (c->ctx->out_ifc_list[c->ifc_idx].bufferswitch == 0) {
+         finish_buffer(c, buffer);
       }
+
+      pthread_mutex_unlock(&c->ctx->out_ifc_list[c->ifc_idx].ifc_mtx);
+      return TRAP_E_OK;
+   } else {
+      /* Not enough space for message, finish current buffer and try to store message into next buffer */
+      finish_buffer(c, buffer);
+      buffer = &c->buffers[c->active_buffer];
+
+      pthread_mutex_unlock(&c->ctx->out_ifc_list[c->ifc_idx].ifc_mtx);
+      goto repeat;
    }
 
-exit:
-   /*
-    * Return to blocking_repeat ONLY when the timeout is TRAP_WAIT.
-    * TRAP_HALFWAIT is handled before.
-    */
-   if ((timeout == TRAP_WAIT) && ((result != TRAP_E_OK) || (c->connected_clients == 0))) {
-      goto blocking_repeat;
+timeout:
+   for (i = 0; i < c->clients_arr_size; i++) {
+      if (c->clients[i].sd > 0 && c->clients[i].assigned_buffer == c->active_buffer) {
+         c->clients[i].timeouts++;
+      }
    }
-   return result;
+   pthread_mutex_unlock(&c->ctx->out_ifc_list[c->ifc_idx].ifc_mtx);
+   return TRAP_E_TIMEOUT;
 }
 
-
 /**
  * \brief Set interface state as terminated.
  * \param[in] priv  pointer to module private data
  */
 void tcpip_sender_terminate(void *priv)
 {
    tcpip_sender_private_t *c = (tcpip_sender_private_t *) priv;
+
+   uint32_t i;
+   uint64_t sum;
+
    if (c != NULL) {
+      do {
+         usleep(10000);
```

Comment:
This usleep should be at least commented, because its reason for being there is not obvious at all. So, what is the point of this sleep?